File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    static ArrayList<Integer> answer = new ArrayList<>();
    static HashMap<String, Integer> bestGenres = new HashMap<>();
    static HashMap<String, List<int[]>> eachSongs = new HashMap<>();
    
    public int[] solution(String[] genres, int[] plays) {
        
        for (int i = 0; i < genres.length; i++) {
            bestGenres.put(genres[i], bestGenres.getOrDefault(genres[i], 0) + plays[i]);
            
            eachSongs.putIfAbsent(genres[i], new ArrayList<>());
            eachSongs.get(genres[i]).add(new int[]{i, plays[i]});
        }
        
        List<String> genreList = new ArrayList<>(bestGenres.keySet());
        genreList.sort((a, b) -> bestGenres.get(b) - bestGenres.get(a));
        
        for (String g : genreList) {
            
            List<int[]> songs = eachSongs.get(g);
            
            songs.sort((a, b) -> {
                       if (a[1] == b[1]) {
                           return a[0] - b[0];
                       }
                       return b[1] - a[1];
            });
            
            answer.add(songs.get(0)[0]);
            
            if (songs.size() > 1) {
                answer.add(songs.get(1)[0]);
            }
        }
        
        return answer.stream().mapToInt(i->i).toArray();
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    let map = new Map();
    
    genres.forEach((genre, index) => {
        if (!map.has(genre)) {
            map.set(genre, {total : 0, songs: []});
        }
        const gInfo = map.get(genre);
        gInfo.total += plays[index];
        gInfo.songs.push({ index, play: plays[index] });
    })
    
    const sortedMap = [...map.entries()].sort((a, b) => b[1].total - a[1].total);
    
    sortedMap.forEach(([_, gInfo]) => {
        gInfo.songs
            .sort((a, b) => {
                if (a.play == b.play) {
                    return a.index - b.index;
                }
                return b.play - a.play;
            })
            .slice(0, 2)
            .forEach(song => answer.push(song.index));
    })
    
    return answer;
}